lib/Dialect/FIRRTL/FIRRTLFolds.cpp

```diff
@@ -2256,6 +2256,77 @@ struct FoldResetMux : public mlir::RewritePattern {
 };
 } // namespace
 
+namespace {
+/// This canonicalizer provides the following patterns:
+///   reset(reg) = 0 ==> connect(reg, and(reg, x)) ==> reg -> 0
+///   reset(reg) = 0 ==> connect(reg, and(x, reg)) ==> reg -> 0
+///   reset(reg) = 1 ==> connect(reg, or(reg, x))  ==> reg -> 1
+///   reset(reg) = 1 ==> connect(reg, or(x, reg))  ==> reg -> 1
+///
+/// Justification: The initial value of a register is indeterminant, which means
+/// we are free to choose any initial value when optimizing the circuit. For the
+/// AND patterns, if the reset is zero, and we assume the initial value is zero,
+/// then the register will always be zero. For the OR patterns, if the reset is
+/// one, and we assume the initial value is one, then the register will always
+/// be one.
+struct RegResetAndOrOfSelf : public mlir::OpRewritePattern<RegResetOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(RegResetOp op,
+                                PatternRewriter &rewriter) const override {
+    // Do not fold the register away if it is important.
+    if (hasDontTouch(op.getOperation()) || !AnnotationSet(op).empty() ||
+        op.isForceable())
+      return failure();
+
+    // This canonicalization only applies when the register holds 1 bit.
+    auto type = dyn_cast<UIntType>(op.getResult().getType());
+    if (!type || type.getWidthOrSentinel() != 1)
+      return failure();
+
+    // This canonicalization only applies when the reset is a constant.
+    auto reset =
+        dyn_cast_or_null<ConstantOp>(op.getResetValue().getDefiningOp());
+    if (!reset)
+      return failure();
```

Comment:
Nit: please be exact. This is the _reset value_ and not the _reset_ (or exactly, the _reset signal_).